### ISSUE TITLE
Audit the remaining manual implementations of ToAnimatedZero

### ISSUE THREAD

**Issue by nox** [servo/servo#18253]
The following types unconditionally return an error in their `ToAnimatedZero` impl:

* `properties::longhands::border_spacing::computed_value::T`
* `properties::longhands::font_stretch::computed_value::T`
* `properties::longhands::font_size_adjust::computed_value::T`
* `properties::longhands::vertical_align::computed_value::T`
* `properties::longhands::visibility::computed_value::T`
* `values::animated::color::Color`
* `values::computed::ClipRect`
* `values::computed::background::BackgroundSize`
* `values::computed::border::BorderCornerRadius`
* `values::computed::length::MaxLength`
* `values::computed::text::LineHeight`
* `values::generics::basic_shape::ShapeSource<B, T, U>`
* `values::generics::text::Spacing<V>`

The following enums return an error for some of their variants, but they don't in their `Animate` and `ComputeSquaredDistance` impls:

* `values::computed::length::LengthOrPercentageOrAuto`
* `values::computed::length::LengthOrPercentageOrNone`

Finally, the impl for `properties::animated_properties::IntermediateSVGPaint` silences errors returned by the `to_animated_zero` call on its `fallback` field and that doesn't look correct to me, given the errors from that field in the `Animate` and `ComputeSquaredDistance` aren't silenced.

Cc @BorisChiou @birtles @emilio

**Comment by BorisChiou** [comment]
cc @hiikezoe

**Comment by hiikezoe** [comment]
Basically we need ToAnimatedZero for properies in nsSMILCSSProperty::IsPropertyAnimatable().
https://hg.mozilla.org/mozilla-central/file/d10c97627b51/dom/smil/nsSMILCSSProperty.cpp#l172

**Comment by birtles** [comment]
We only need it to not error for additive property types. Specifically float, percent, coord, and color.
http://searchfox.org/mozilla-central/rev/18c16ebf818abb86805ce08a6e537e4cd826f044/dom/smil/nsSMILCSSValueType.cpp#90

**Comment by nox** [comment]
That doesn't explain a lot to me. Why is it a trait at all if only 4 things are supposed to support it? Don't types that transitively use float, percent, coord, and color (whatever these are) also need to support that?

**Comment by nox** [comment]
It also doesn't tell me *what kind* of values they should return. Are they presented to the users? For example, would it be more convenient for the animated zero of a filter list to be an empty filter list, or a list of the animated zeros of each filter?

**Comment by birtles** [comment]
What's the purpose of this issue?

**Comment by nox** [comment]
@birtles Am trying to check the correctness of the few impls listed in the issue. @hiikezoe mentions a list of properties from Gecko's code and I see `font-stretch` in it, but our implementation of `ToAnimatedZero` just returns `Err(())`, so surely that's not correct, right?

If that is indeed wrong, I'll just check according to the list and do the necessary fixes.

**Comment by birtles** [comment]
Yes, that sounds wrong since `font-stretch` should be additive per [SVG](https://www.w3.org/TR/SVG11/text.html#FontStretchProperty).

It looks like Gecko is buggy here. On the one had [it makes it additive](http://searchfox.org/mozilla-central/rev/cd82cacec2cf734768827ff85ba2dba90a534c5e/layout/style/StyleAnimationValue.cpp#2921) but it doesn't suppose a zero value for enumerated types so this property will be additive but unable to be used in _by-animation_ which seems wrong.

**Comment by birtles** [comment]
Basically, `ToAnimatedZero` should be supported for any properties that are _both_ (a) animatable by SMIL and (b) additive, where (b) is obviously a subset of (a). Clearly Gecko has some gaps in this, however.

Also (and I think we have a comment about this somewhere), its only purpose is to support SMIL's _by-animation_ where we say "animate from the underlying value by _x_ amount". Internally that is implemented as an animation from _zero_ → _x_ and then the result is added to the underlying value. The requirement is that when we add _zero_ to the underlying value we get the underlying value, and also that _zero_ is something that interpolates with _x_ (hence why we pass _x_ into `ToAnimatedZero`). The value is never shown to the user.

**Comment by nox** [comment]
@birtles Thanks for the explanations, I'll go through the list and use your links to try to find out what is done in Gecko.
